src/services/redisSessionStorage.ts

```diff
@@ -2,64 +2,179 @@
+  timestampEpochMillis: IntegerFromString,
+  user: User
+});
+type Session = t.TypeOf<typeof Session>;
+
+export interface ISessionState {
+  readonly expired: boolean;
+  readonly expireAt?: number;
+  readonly newToken?: string;
+  readonly user?: User;
+}
 
 export default class RedisSessionStorage implements ISessionStorage {
   constructor(
     private readonly redisClient: redis.RedisClient,
-    private readonly tokenDuration: number
+    private readonly tokenDurationSecs: number,
+    private readonly tokenService: TokenService
   ) {}
 
   /**
    * {@inheritDoc}
    */
-  public set(token: string, user: User): void {
-    // Set key to hold the string value. This data is set to expire (EX) after
-    // `this.tokenDuration` seconds.
-    // @see https://redis.io/commands/set
-    this.redisClient.set(token, JSON.stringify(user), "EX", this.tokenDuration);
+  public set(
+    token: string,
+    user: User,
+    timestampEpochMillis: number
+  ): Promise<Either<Error, boolean>> {
+    return new Promise(resolve => {
+      // Set a key to hold the fields value.
+      // @see https://redis.io/commands/hmset
+      this.redisClient.hmset(
+        token,
+        {
+          timestampEpochMillis,
+          user: JSON.stringify(user)
+        },
+        (err, response) => {
+          if (err) {
+            return resolve(left<Error, boolean>(err));
+          }
+
+          resolve(right<Error, boolean>(response));
+        }
+      );
+    });
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public async get(token: string): Promise<Either<Error, ISessionState>> {
+    const errorOrSession = await this.getSession(token);
+
+    if (isLeft(errorOrSession)) {
+      const error = errorOrSession.value;
+      return left(error);
+    }
+
+    const session = errorOrSession.value;
+    const user = session.user;
+
+    // Check if the token has expired. We don't remove expired token
+    // because a client can later refresh the session.
+    const expire =
+      (session.timestampEpochMillis as number) + this.tokenDurationSecs * 1000;
+    if (expire < Date.now()) {
+      return left<Error, ISessionState>(new Error("Token has expired"));
+    }
+
+    return right<Error, ISessionState>({
+      expireAt: expire,
+      expired: false,
+      user
+    });
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public async refresh(token: string): Promise<Either<Error, ISessionState>> {
+    const errorOrSession = await this.getSession(token);
+
+    if (isLeft(errorOrSession)) {
+      const error = errorOrSession.value;
+      return left(error);
+    }
+
+    const session = errorOrSession.value;
+    const user = session.user;
+
+    // Compute a new token.
+    const newToken = this.tokenService.getNewToken();
+
+    // Set a session with the new token, delete the old one.
+    const timestamp = Date.now();
+    return Promise.all([this.set(newToken, user, timestamp), this.del(token)])
+      .then(() =>
+        right<Error, ISessionState>({
```

Comment:
this looks incorrect, `Promise.all([this.set(newToken, user, timestamp), this.del(token)])` will resolve to `[Either<Error, boolean>, Either<Error, boolean>]` - you should check the result value - also you should `await` the promise for consistency with the rest of the code:

```ts
const [ setResult, delResult ] = await Promise.all([this.set(newToken, user, timestamp), this.del(token)]);
if(isLeft(setResult.alt(delResult))) {
  return left(...)
}
return right(...)
```